### ISSUE TITLE
Ouroboros Close don't return because deadlock

### ISSUE THREAD

**Issue by safanaj** [blinklabs-io/gouroboros#257]
With the introduction of waitGroup now there are 2 deadlock during the Close method.

The first encountered is because the muxer.Stop don't return until the connection still open, the muxer has the readLoop goroutine blocked on [binary.Read](https://github.com/blinklabs-io/gouroboros/blob/ea94cd02e649dc22e2f0dd957c05149dfbbcaf2a/muxer/muxer.go#L190) waiting for readable data on the connection, to avoid to send undesired error the muxer done channel have to be closed before the connection.

The second deadlock is caused by the [mini-protocol error forwarder goroutine](https://github.com/blinklabs-io/gouroboros/blob/ea94cd02e649dc22e2f0dd957c05149dfbbcaf2a/ouroboros.go#L299) , this prevent the waitGroup to complete, so the [protoErrorChan](https://github.com/blinklabs-io/gouroboros/blob/ea94cd02e649dc22e2f0dd957c05149dfbbcaf2a/ouroboros.go#L159) has to be closed before to wait for the waitGroup.

Proposed patch:
```
diff --git a/muxer/muxer.go b/muxer/muxer.go
index 75c8f12..d18911a 100644
--- a/muxer/muxer.go
+++ b/muxer/muxer.go
@@ -46,6 +46,7 @@ const (
 type Muxer struct {
 	errorChan         chan error
 	conn              net.Conn
+	connp             *net.Conn
 	sendMutex         sync.Mutex
 	startChan         chan bool
 	doneChan          chan bool
@@ -57,9 +58,10 @@ type Muxer struct {
 }
 
 // New creates a new Muxer object and starts the read loop
-func New(conn net.Conn) *Muxer {
+func New(connp *net.Conn) *Muxer {
 	m := &Muxer{
-		conn:              conn,
+		conn:              *connp,
+		connp:             connp,
 		startChan:         make(chan bool, 1),
 		doneChan:          make(chan bool),
 		errorChan:         make(chan error, 10),
@@ -81,7 +83,7 @@ func (m *Muxer) Start() {
 }
 
 // Stop shuts down the muxer
-func (m *Muxer) Stop() {
+func (m *Muxer) Stop() error {
 	// We use a mutex to prevent this function from being called multiple times
 	// concurrently, which would cause a race condition
 	m.stopMutex.Lock()
@@ -89,11 +91,24 @@ func (m *Muxer) Stop() {
 	// Immediately return if we're already shutting down
 	select {
 	case <-m.doneChan:
-		return
+		return nil
 	default:
 	}
 	// Close doneChan to signify that we're shutting down
 	close(m.doneChan)
+	// Close connection an set it to nil
+	if m.conn != nil {
+		if err := m.conn.Close(); err != nil {
+			// the muxer Stop is called only by ouroborus Close,
+			// we cannot use the muxer error chan b/c would call ouroborus Close.
+			// ouroborus Close will forward this error to the errorChan
+			return err
+		}
+		if m.connp != nil {
+			// this prevents double close of connection
+			*m.connp = nil
+		}
+	}
 	// Wait for other goroutines to shutdown
 	m.waitGroup.Wait()
 	// Close protocol receive channels
@@ -103,6 +118,7 @@ func (m *Muxer) Stop() {
 	}
 	// Close ErrorChan to signify to consumer that we're shutting down
 	close(m.errorChan)
+	return nil
 }
 
 // SetDiffusionMode sets the muxer diffusion mode after the handshake completes
diff --git a/ouroboros.go b/ouroboros.go
index e3cbcbf..0b82e1a 100644
--- a/ouroboros.go
+++ b/ouroboros.go
@@ -144,19 +144,23 @@ func (o *Ouroboros) Close() error {
 	close(o.doneChan)
 	// Gracefully stop the muxer
 	if o.muxer != nil {
-		o.muxer.Stop()
+		if err := o.muxer.Stop(); err != nil {
+			o.errorChan <- err
+		}
 	}
-	// Close the underlying connection
+	// Close the underlying connection,
+	// the muxer already closed it
 	if o.conn != nil {
 		if err := o.conn.Close(); err != nil {
 			return err
 		}
 	}
+	// Terminate mini-protocols error forwarder goroutine
+	close(o.protoErrorChan)
 	// Wait for other goroutines to finish
 	o.waitGroup.Wait()
-	// Close channels
+	// Close error channel
 	close(o.errorChan)
-	close(o.protoErrorChan)
 	// We can only close a channel once, so we have to jump through a few hoops
 	select {
 	// The channel is either closed or has an item pending
@@ -211,7 +215,7 @@ func (o *Ouroboros) TxSubmission() *txsubmission.TxSubmission {
 // setupConnection establishes the muxer, configures and starts the handshake process, and initializes
 // the appropriate mini-protocols
 func (o *Ouroboros) setupConnection() error {
-	o.muxer = muxer.New(o.conn)
+	o.muxer = muxer.New(&o.conn)
 	// Start Goroutine to pass along errors from the muxer
 	o.waitGroup.Add(1)
 	go func() {
```
The solution for the muxer.Stop deadlock is a bit ugly, passing a net.Conn pointer just to prevent double close (very unlikely to happen). A clearer solution should just let only the muxer to manage the connection (as it is the "network guy" as I know), but the `WithConnection` option looks like a fundamental part of the lib api, so the `WithConnection` is the only reason to keep the attribute on the Ouroborus struct. In case of error during the connection close the muxer will return immediately and the Ourobourus Close can forward it to the error channel to notify user.

**Comment by safanaj** [comment]
I open the #258 that would apply the patch attached to this issue

**Comment by agaffney** [comment]
How are you encountering this deadlock on `.Close()`? Are you creating a new `Ouroboros` object without providing a `net.Conn` object and then immediately calling `.Close()`?

**Comment by safanaj** [comment]
I will post here a reproducible code, but basically just do:
1. Create an ouroborus with connection.
2. Start chainsync and get a tip
3. Close the Ouroboros

**Comment by safanaj** [comment]
Code to reproduce:
```
func main() {
	ctx := context.Background()
	address := "<relay:port>"
	conn, err := (&net.Dialer{Timeout: time.Second}).DialContext(ctx, "tcp", address)
	if err != nil {
		panic(err)
	}

	o, _ := ouroboros.New(
		ouroboros.WithConnection(conn),
		ouroboros.WithNetwork(ouroboros.NetworkByName("mainnet")),
		ouroboros.WithNodeToNode(true),
		ouroboros.WithKeepAlive(true),
	)

	// die on error
	go func() {
		err := <-o.ErrorChan()
		if err != nil {
			fmt.Printf("ERROR from ouroboros: %v", err)
		}
	}()

	o.ChainSync().Client.Start()
	tip, err := o.ChainSync().Client.GetCurrentTip()
	if err != nil {
		o.Close()
		panic(err)
	}
	fmt.Printf("node: %s tip: slot = %d, hash = %x\n", address, tip.Point.Slot, tip.Point.Hash)

	o.Close()
	fmt.Println("o.Close() terminated")
	select {}
}
```
The "o.Close() terminated" is never printed, below output before and after the patch applied:
```
$ go mod tidy && go mod vendor && go run main.go 
node: <relay:port redacted> tip: slot = 92419649, hash = f5ddeae422e8a2f2722019e193fe89011741a438c5091796e661dfb699e09cc5
^Csignal: interrupt
$ go mod edit -replace=github.com/blinklabs-io/gouroboros@v0.38.4=github.com/safanaj/gouroboros@v0.38.4.1 && go mod tidy && go mod vendor && go run main.go 
node: <relay:port redacted> tip: slot = 92419704, hash = 5cdefabfa75034b3230feaf913902ddc375d09cb013aa0093ccec01a0f088279
o.Close() terminated
^Csignal: interrupt
```

**Comment by safanaj** [comment]
This happens of course with any protocol, the connection I am using is n2n, for example just sending a ping, so replacing:
```
	o.ChainSync().Client.Start()
	tip, err := o.ChainSync().Client.GetCurrentTip()
	if err != nil {
		o.Close()
		panic(err)
	}
	fmt.Printf("node: %s tip: slot = %d, hash = %x\n", address, tip.Point.Slot, tip.Point.Hash)
```
with:
```
o.KeepAlive().Client.SendMessage(keepalive.NewMsgKeepAlive(10))
```
the result is the same, we never see the `o.Close() terminated` print.

Just playing with keepalive protocol in n2n, I just noticed that it panics with `close of closed channel` if the protocol is started, just calling `o.KeepAlive().Client.Start()` , I am not sure but I guess is not the desired behavior. I fill another issue after investigation.

**Comment by safanaj** [comment]
The keepalive thing is probably because it is already started, so the explicit client start call cause the panic, eventually some implicit/hidden protection for that would be good, like a sync.Once in the keepalive client start to prevent double start

**Comment by agaffney** [comment]
This is at least partially fixed by https://github.com/blinklabs-io/gouroboros/pull/263

**Comment by safanaj** [comment]
@agaffney I am not sure, as I did not test this yet, but as I quickly see the muxer.Stop would still not returning and so it is blocking the ouroboros Close.
I think muxer Stop is not returning because the readLoop can be blocked on the net.Conn ([here](https://github.com/blinklabs-io/gouroboros/blob/f8938aa7d29331fb42fd8f21a4cfa779be485c60/muxer/muxer.go#L191)) on the binary.Read and still blocked until the net.Conn is closed, doing so it is causing a circular waiting until forever.

Not sure if I am clear here.

**Comment by agaffney** [comment]
> @agaffney I am not sure, as I did not test this yet, but as I quickly see the muxer.Stop would still not returning and so it is blocking the ouroboros Close. I think muxer Stop is not returning because the readLoop can be blocked on the net.Conn ([here](https://github.com/blinklabs-io/gouroboros/blob/f8938aa7d29331fb42fd8f21a4cfa779be485c60/muxer/muxer.go#L191)) on the binary.Read and still blocked until the net.Conn is closed, doing so it is causing a circular waiting until forever.
> 
> Not sure if I am clear here.

I'm pretty sure that you're right. This problem has existed for a long time, but it's only just now been exposed by the addition of `sync.WaitGroup` in the muxer.
